File: cpp/SelectionSort.cpp
#include <ctime>
#include <fstream>
#include <iostream>
#include <string>
#include <vector>
using namespace std;

int main(int argc, char* argv[]) {
    int size;
    // Get command line input
    if (argc > 1) {
        size = stoi(argv[1]);
    } else {
        size = 10000;
    }
    cout << "Looking for size: " << size << endl;
    
    // Declare a vector of longs to store the numbers
    vector<long> numbers;
    
    // Read size numbers from numbers.txt
    ifstream inFile;
    inFile.open("numbers.txt");
    long num;
    for(int i=0; i<size; ++i) {
        if(inFile) {
            inFile >> num;
            numbers.push_back(num);
        }
    }

    // Print the vector size (to make sure it matches the size printed above)
    cout << "The vector has a size of " << numbers.size() << endl;

    //Selection Sort the vector
    int swapIndex, i, minIndex;
    long temp;
    for (swapIndex = 0; swapIndex < numbers.size()-1; ++swapIndex) {
        // Loop through vector starting at swapIndex and keep track of min
        minIndex = swapIndex;
        for (i = swapIndex+1; i < numbers.size(); ++i) {
            if (numbers[i] < numbers[minIndex]) {
                // We have a new minimum
                minIndex = i;
            }
        }
        // Swap min value into swapIndex spot in vector
        temp = numbers[swapIndex];
        numbers[swapIndex] = numbers[minIndex];
        numbers[minIndex] = temp;
    }

    // print the first and last ten numbers from the vector to the console
    for(int i = 0; i < 10; i++) {
        cout << numbers[i] << ", ";
    }
    for(int j = numbers.size() - 10; j < numbers.size(); j++) {
        cout << numbers[j] << ", ";
    }
    return 0;
}

File: javascript/SelectionSort.js

var size;
//Get command line input
var arguments = process.argv; 
if(arguments.length > 0) {
	size = parseInt(arguments[2]);
} else {
	size = 10000;
}
console.log("Looking for size: ", size);

//Create an array to store the numbers
var numbers = [];

//Code taken from 
const { open } = require('node:fs/promises');

//A function to get size numbers from the file
const getNums = async (numbers) => {
	try {
		let i = 0;

		const file = await open('numbers.txt');
	
		for await (const line of file.readLines()) {
			if (i < size) {
				numbers.push(line);
				i ++;
			}
		}
		return numbers;
	}
	catch (e) {
		//Display the error
		console.log(e);
	}	
}

//A function to print the first 10 and last 10 numbers from the file
const print = async (numbers) => {
	try {
		// Print the first and last ten numbers from the vector to the console
		for (let i = 0; i < 10; i ++) {
			console.log(numbers[i]);
		}
		console.log("...")
		for (let j = numbers.length - 10; j < numbers.length; j++) {
			console.log(numbers[j]);
		}
	}
	catch (e) {
		console.log (e);
	}
}

//A function to bubble sort the numbers
const sort = async (numbers) => {
	try {
		// Get the numbers from the file
		var nums = await getNums(numbers);
	
		//Selection sort algorithm 
		var i, swapIndex, min;
		var temp;

		for(swapIndex = 0; swapIndex < nums.length - 1; swapIndex++) {
			//Loop through the array starting at swapIndex and keeps track of the minimum
			min = await swapIndex;
			for (i = swapIndex + 1; i < nums.length; i++) {
				if(parseInt(nums[i]) < parseInt(nums[min])) {
					//We have a new minimum
					min = await i;
				}
			}
			//Swap min value into swapIndex spot in vector
			temp = await nums[swapIndex];
			nums[swapIndex] = await nums[min];
			nums[min] = await temp;
		}
		
		await print(nums);

		return nums;
	}
	catch (e) {
		console.log(e);
	}
}

// Calls the sort method
sort(numbers);